namer/rancher/src/main/scala/io/buoyant/namer/rancher/RancherClient.scala

```diff
@@ -0,0 +1,149 @@
+  objectMapper.registerModule(DefaultScalaModule)
+  objectMapper.registerModule(customTypes)
+  objectMapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE)
+
+  def apply() = objectMapper
+}
+
+class RancherClient(
+  maxWait: Int,
+  log: Logger,
+  params: Stack.Params,
+  stats: StatsReceiver
+)(implicit val timer: Timer) {
+  private[this] val numVersions = stats.counter("metadata_versions")
+
+  private[this] val service = Http.client
+    .withStack(Http.client.stack.remove(Retries.Role))
+    .withParams(Http.client.params ++ params)
+    .withLabel("io.buoyant.namer.rancher.client")
+    .withTracer(NullTracer)
+    .newService(s"/$$/inet/rancher-metadata/80")
+
+  private[this] val version: Activity[String] =
+    Activity(Var.async[State[String]](Activity.Pending) { state =>
+      val done = new AtomicBoolean(false)
+      @volatile var v: Option[String] = None
+
+      Future.whileDo(!done.get) {
+        log.trace("starting version request")
+
+        val request = http.Request(http.Method.Get, s"/2015-12-19/version?wait=true&maxWait=${maxWait}&value=${v}")
+        request.host = "rancher-metadata"
+
+        service(request)
+          .map { _.getContentString().trim }
+          .transform {
+            case Return(newVersion) if !v.contains(newVersion) =>
+              log.info("fetched new metadata-version: %s", newVersion)
+              v = Some(newVersion)
+              numVersions.incr()
+              state.update(Activity.Ok(newVersion))
+              Future.Unit
+            case Return(newVersion) if v.contains(newVersion) =>
+              log.trace("metadata-version didn't change. Old: %s, received: %s", v, newVersion)
+              Future.Unit
+            case Throw(e) =>
+              // TODO: This should probably be handled by a exponentail back-off
+              // in the Finagle client - but I'm not sure how to achieve that
+              log.error("failed to fetch metadata-version. %s", e)
+              if (!v.isEmpty) {
+                // We have previously managed to get a version, but now we can't
+                // so we set mark the activity as failed
+                state.update(Activity.Failed(e))
+              }
+              Future.sleep(Duration.fromFractionalSeconds(5.0))
+          }
+      }
+
+      Closable.make { _ =>
+        done.set(true)
+        Future.Unit
+      }
+    })
+
+  private[this] val containers: Activity[List[RancherContainer]] = version.flatMap { version =>
+    Activity.future {
+      val request = http.Request(http.Method.Get, "/2015-12-19/containers")
+      request.host = "rancher-metadata"
+      request.accept = "application/json"
+
+      service(request)
+        .map { _.getContentString() }
+        .map { RancherParser().readValue[List[RancherContainer]] }
+        .onSuccess { containers =>
+          log.debug("received %s new containers", containers.size)
+        }: Future[List[RancherContainer]]
```

Comment:
you shouldn't nee this type annotation

Reply:
You are very right. While trying to figure out how stuff worked, I found it useful to add in extra typo annotations to make sure my assumptions about types held up...